### ISSUE TITLE
When to update the improved version

### ISSUE THREAD

**Issue by KKDDD** [cfzd/Ultra-Fast-Lane-Detection#260]
Is it convenient to disclose how it was improved?

**Comment by KKDDD** [comment]
@cfzd 
 Is the hybrid anchor carried out on different feature scales? Is it convenient to reveal his more detailed description?

**Comment by KKDDD** [comment]
@cfzd 
Do you mean that lane line detection is now carried out from both horizontal and vertical directions?

**Comment by cfzd** [comment]
@KKDDD 
Not both. For each lane, only one anchor is adopted. But anchors can be different regarding the lane's orientation.

**Comment by cfzd** [comment]
@KKDDD 
Sorry, I forget to tell you that we split the classification of localization and existence. If the existence and localization parts (i.e., `x[:,:grid-1,:,:]` and `x[:,:-1,:,:]`) are trained together, then the above loss has no influence. You should do something like this:

```
loc, ext = network(data)
loc = ce_loss(loc, loc_label)
ext = ce_loss(ext, ext_label)

loc_real = l1_loss( (loc.softmax() * k).sum(), loc_real_label )
```

**Comment by KKDDD** [comment]
I don't understand what you mean too much. Can you give me a little more detailed explanation?
Are `loc `and `ext ` the output of the classification and segmentation in the original network structure, or the output of the new network structure?

**Comment by cfzd** [comment]
@KKDDD 
In the previous version, the classification of "no lane" is trained together like this:
```
loss = ce_loss(output, target)
# output[:, :-1, :, :] is the "no lane case"
```

In the improved version, there are separated:
```
loss1 = ce_loss(output[:,:grid-1,:,:], loc_target)
loss2  = ce_loss(output[:, :-1, :, :], ext_target)
```

**Comment by ma-cg** [comment]
When will you update the new version? I've been waiting for a long time

**Comment by KKDDD** [comment]
I tried to train the positioning and existence separately, and added a new loss of the final desired position,
`cls, ext = net(input)`
`cls_loss = ce_loss(loc, loc_label)` 
`ext_loss = ce_loss(ext, ext_label)`
`loc_loss = l1_loss(cls.softmax()*k.sum() , loc_label)` 
Among them, for rows without lane points, the calculation of positioning (loc_loss) and classification loss (cls_loss) is ignored.

However, the accuracy rate is not improved much, and it is best to only achieve an accuracy rate of 70% (0.3% increase). Does this new output structure and positioning loss function improve the accuracy of the new version of the model?

**Comment by cfzd** [comment]
@KKDDD 
Yes, in our ablation, the `loc_loss` brings about a 0.5% performance gain. The biggest improvement is from the hybrid anchor, which contains both row and column anchors.

**Comment by KKDDD** [comment]
@cfzd 
Is your 0.5% accuracy improvement still the original parameter configuration used?
Because I found that the accuracy improvement is very unstable after the positioning is separated。

**Comment by cfzd** [comment]
@KKDDD 
We add a small coefficient for the `loc_loss`, like 0.05. Could this be the problem?

**Comment by KKDDD** [comment]
yes， I want to know whether other parameters, such as `row_anchor` and `grid_num,` and so on are the same as the value of the current warehouse.

**Comment by KKDDD** [comment]
@cfzd 
yes， I want to know whether other parameters, such as row_anchor and grid_num, and so on are the same as the value of the current warehouse.
